File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/storage/StorageKey.java
package org.whispersystems.signalservice.api.storage;

import org.whispersystems.signalservice.api.kbs.MasterKey;
import org.whispersystems.signalservice.internal.util.Hex;
import org.whispersystems.util.Base64;
import org.whispersystems.util.StringUtil;

import java.util.Arrays;

import static org.whispersystems.signalservice.api.crypto.CryptoUtil.hmacSha256;

/**
 * Key used to encrypt data on the storage service. Not used directly -- instead we used keys that
 * are derived for each item we're storing.
 *
 * Created via {@link MasterKey#deriveStorageServiceKey()}.
 */
public final class StorageKey {

  private static final int LENGTH = 32;

  private final byte[] key;

  public StorageKey(byte[] key) {
    if (key.length != LENGTH) throw new AssertionError();

    this.key = key;
  }

  public StorageManifestKey deriveManifestKey(long version) {
    return new StorageManifestKey(derive("Manifest_" + version));
  }

  public StorageItemKey deriveItemKey(byte[] key) {
    return new StorageItemKey(derive("Item_" + Base64.encodeBytes(key)));
  }

  private byte[] derive(String keyName) {
    return hmacSha256(key, StringUtil.utf8(keyName));
  }

  public byte[] serialize() {
    return key.clone();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || o.getClass() != getClass()) return false;

    return Arrays.equals(((StorageKey) o).key, key);
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(key);
  }
}

File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/storage/StorageKey.kt
package org.whispersystems.signalservice.api.storage

import org.signal.core.util.Base64.encodeWithPadding
import org.signal.core.util.CryptoUtil
import org.whispersystems.signalservice.api.kbs.MasterKey
import org.whispersystems.util.StringUtil

/**
 * Key used to encrypt data on the storage service. Not used directly -- instead we used keys that
 * are derived for each item we're storing.
 *
 * Created via [MasterKey.deriveStorageServiceKey].
 */
class StorageKey(val key: ByteArray) {
  init {
    check(key.size == 32)
  }

  fun deriveManifestKey(version: Long): StorageManifestKey {
    return StorageManifestKey(derive("Manifest_$version"))
  }

  fun deriveItemKey(rawId: ByteArray): StorageItemKey {
    return StorageItemKey(derive("Item_" + encodeWithPadding(rawId)))
  }

  private fun derive(keyName: String): ByteArray {
    return CryptoUtil.hmacSha256(key, StringUtil.utf8(keyName))
  }

  fun serialize(): ByteArray {
    return key.clone()
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as StorageKey

    return key.contentEquals(other.key)
  }

  override fun hashCode(): Int {
    return key.contentHashCode()
  }
}
